pkg/verification/verify.go

```diff
@@ -18,17 +18,196 @@ package verification
+	}
+
+	return nil
+}
+
+// Verify the leaf certificate's subject Common Name name matches a provided Common Name
+func verifySubjectCommonName(cert *x509.Certificate, opts VerifyOpts) error {
+	if opts.CommonName == "" {
+		return nil
+	}
+
+	if cert.Subject.CommonName != opts.CommonName {
+		return fmt.Errorf("the certificate's subject Common Name %s does not match the provided Common Name %s", cert.Subject.CommonName, opts.CommonName)
+	}
+	return nil
+}
+
+// If embedded in the TSR, verify the TSR's leaf certificate matches a provided TSA certificate
+func verifyEmbeddedLeafCert(tsaCert *x509.Certificate, opts VerifyOpts) error {
+	if opts.TSACertificate != nil && !opts.TSACertificate.Equal(tsaCert) {
+		return fmt.Errorf("certificate embedded in the TSR does not match the provided TSA certificate")
+	}
+	return nil
+}
+
+// Verify the leaf's EKU is set to critical, per RFC 3161 2.3
+func verifyLeafCertEKU(cert *x509.Certificate) error {
+	var criticalEKU bool
+	for _, ext := range cert.Extensions {
+		if ext.Id.Equal(EKUOID) {
+			criticalEKU = ext.Critical
+		}
+	}
+	if !criticalEKU {
+		return errors.New("certificate must set EKU to critical")
+	}
+	return nil
+}
+
+func verifyLeafCert(ts timestamp.Timestamp, opts VerifyOpts) error {
+	if len(ts.Certificates) == 0 && opts.TSACertificate == nil {
+		return fmt.Errorf("leaf certificate must be present the in TSR or as a verify option")
+	}
+
+	errMsg := "failed to verify leaf cert"
+
+	var leafCert *x509.Certificate
+	if len(ts.Certificates) != 0 {
+		leafCert = ts.Certificates[0]
+		if opts.TSACertificate != nil && !leafCert.Equal(opts.TSACertificate) {
+			return fmt.Errorf("leaf certificate included in the TSR does not match the one provided as a verify option")
+		}
+
+		err := verifyEmbeddedLeafCert(leafCert, opts)
+		if err != nil {
+			return fmt.Errorf("%s: %w", errMsg, err)
+		}
+	} else {
+		leafCert = opts.TSACertificate
+	}
+
+	err := verifyLeafCertEKU(leafCert)
+	if err != nil {
+		return fmt.Errorf("%s: %w", errMsg, err)
+	}
+
+	err = verifyESSCertID(leafCert, opts)
+	if err != nil {
+		return fmt.Errorf("%s: %w", errMsg, err)
+	}
+
+	err = verifySubjectCommonName(leafCert, opts)
+	if err != nil {
+		return fmt.Errorf("%s: %w", errMsg, err)
+	}
+
+	return nil
+}
+
+func verifyExtendedKeyUsage(cert *x509.Certificate) error {
+	certEKULen := len(cert.ExtKeyUsage)
+	if certEKULen != 1 {
+		return fmt.Errorf("cert has %d extended key usages, expected only one", certEKULen)
+	}
+
+	if cert.ExtKeyUsage[0] != x509.ExtKeyUsageTimeStamping {
+		return fmt.Errorf("leaf cert EKU is not set to TimeStamping as required")
+	}
+	return nil
+}
+
+// Verify the TSA certificate and the intermediates (called "EKU chaining") all
+// have the extended key usage set to only time stamping usage
+func verifyLeafAndIntermediatesEKU(opts VerifyOpts) error {
+	if opts.TSACertificate == nil || opts.Intermediates == nil {
+		return nil
+	}
+	leafCert := opts.TSACertificate
```

Comment:
This needs to account for the leaf certificate coming from either the options or embedded in the timestamp.

Reply:
I'm going remove this function and add the leaf certificate check to the larger leaf certificate verification function. The intermediate verification will be done on its own.